File: codes/chapitre05/fiches_patients.py
from numpy import array


def remplir():
    np = 0
    while not 3 <= np <= 100:
        np = int(input("Entrer le nombre de patients [3, 100] ? "))

    noms = array([str]*np)
    for i in range(np):
        noms[i] = input(f"Nom patient {i+1} ? ")
        while recherche(noms, i-1, noms[i]) != -1:
            print(f"{noms[i]} existe déjà!")
            noms[i] = input(f"Nom patient {i+1} ? ")
    return np, noms


def recherche(noms, np, nom):
    i = 0
    p = -1
    nom = nom.upper()
    while i < np and p == -1:
        noms_m = noms[i].upper()
        if noms_m == nom:
            p = i
        else:
            i = i + 1
    return p


# PP
np, noms = remplir()
nom_patient = ""
while nom_patient.upper() != "FIN":
    nom_patient = input("Patient à rechercher ? ")
    nf = recherche(noms, np, nom_patient)
    if nf == -1:
        print(f"Le patient est introuvable!")
    else:
        print(f"Fiche n°{nf+1}")

File: codes/chapitre05/backup/fiches_patients.pas
program fiches_patients;
type
  tab = array [1..100] of string;

Function Recherche(t:tab;n:integer;v:string):integer;
Var
  i, p: integer;
Begin
  p := 0; i := 1;
  While (i <= n) and (p = 0) Do
  Begin
    if t[i] = v then p := i else i := i + 1;
  End;
  Recherche := p;
End;

Procedure Remplir(var t:tab; var n:integer);
Var
  i: integer;
Begin
  Repeat
    Write('Donner le Nbre de Patients ? ');
    Readln(n);
  until (3 <= n) and (n <= 100);
  for i := 1 to n do
    Repeat
      Write('Patient n°', i, ' ? ');
      Readln(t[i]);
    Until Recherche(t, i-1, t[i]) = 0;
End;

var
  noms: tab;
  nom_patient: string;
  np, nf: integer;
begin
  Remplir(noms, np);
  Repeat
    Write('Patient à rechercher ? ');
    Readln(nom_patient);
    nf := Recherche(noms, np, nom_patient);
    if nf = 0 Then
      Writeln('Le patient est introuvable')
    else
      Wrireln('Fiche n°', nf);
  Until Upcase(nom_patient) = 'FIN';
  Readln;
end.

